File: experiments/tools/connectbot/compose/fragment_host_editor.kt
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.paint
import androidx.compose.ui.graphics.Color
import androidx.compose.foundation.Image
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.sp
import androidx.compose.ui.unit.dp

@Composable
fun fragment_host_editor() {
  Column(modifier = Modifier.fillMaxWidth().fillMaxHeight().verticalScroll(rememberScrollState())) {
    Column(modifier = Modifier.fillMaxWidth().wrapContentHeight()) {
      Box(modifier = Modifier.fillMaxWidth().wrapContentHeight()) {
        Box(modifier = Modifier.align(Alignment.CenterStart)) {
          Box(modifier = Modifier.size(24.0.dp, 24.0.dp).padding(start = 16.0.dp)) { /* TODO Image */ }
        }
        Box(modifier = Modifier.align(Alignment.TopStart)) {
          Text(text = "Protocol", modifier = Modifier.wrapContentWidth().wrapContentHeight().padding(top = 20.0.dp).padding(start = 72.0.dp), color = Color.Unspecified, fontSize = 16.sp)
        }
        Box(modifier = Modifier.align(Alignment.BottomStart)) {
          Text(text = "ssh", modifier = Modifier.wrapContentWidth().wrapContentHeight().padding(bottom = 20.0.dp).padding(start = 72.0.dp).padding(top = 36.0.dp), color = Color.Unspecified, fontSize = 14.sp)
        }
      }
      Row(modifier = Modifier.fillMaxWidth().wrapContentHeight().padding(start = 68.0.dp)) {
        Column(modifier = Modifier.wrapContentHeight().weight(1.0f)) {
          
          OutlinedTextField(value = "", onValueChange = {}, modifier = Modifier.wrapContentHeight().weight(1.0f))
        }
        Box(modifier = Modifier.fillMaxHeight(), contentAlignment = Alignment.CenterStart) {
        IconButton(onClick = {}, modifier = Modifier.background(Color(0x00000000)).size(48.0.dp, 48.0.dp).padding(16.0.dp)) {
          Image(painter = painterResource(id = R.drawable.ic_expand_more), contentDescription = "Expand", contentScale = ContentScale.Fit)
        }
        }
      }
      Box(modifier = Modifier.fillMaxWidth().wrapContentHeight()) {
        Box(modifier = Modifier.align(Alignment.CenterStart)) {
          Box(modifier = Modifier.size(24.0.dp, 24.0.dp).padding(start = 16.0.dp)) { /* TODO Image */ }
        }
        Column(modifier = Modifier.fillMaxWidth().wrapContentHeight().padding(start = 68.0.dp).padding(end = 16.0.dp)) {
          
          OutlinedTextField(value = "", onValueChange = {}, placeholder = { Text("\"Nickname\"") }, modifier = Modifier.fillMaxWidth().wrapContentHeight())
        }
      }
      Box(modifier = Modifier.fillMaxWidth().wrapContentHeight()) {
        Box(modifier = Modifier.align(Alignment.CenterStart)) {
          Box(modifier = Modifier.size(24.0.dp, 24.0.dp).padding(start = 16.0.dp)) { /* TODO Image */ }
        }
        Box(modifier = Modifier.align(Alignment.TopStart)) {
          Text(text = "\"Color category\"", modifier = Modifier.wrapContentWidth().wrapContentHeight().padding(top = 20.0.dp).padding(start = 72.0.dp), color = Color.Unspecified, fontSize = 16.sp)
        }
        Box(modifier = Modifier.align(Alignment.BottomStart)) {
          Text(text = "gray", modifier = Modifier.wrapContentWidth().wrapContentHeight().padding(bottom = 20.0.dp).padding(start = 72.0.dp).padding(top = 36.0.dp), color = Color.Unspecified, fontSize = 14.sp)
        }
      }
      Box(modifier = Modifier.fillMaxWidth().wrapContentHeight()) {
        Box(modifier = Modifier.align(Alignment.CenterStart)) {
          Box(modifier = Modifier.size(24.0.dp, 24.0.dp).padding(start = 16.0.dp)) { /* TODO Image */ }
        }
        Box(modifier = Modifier.align(Alignment.TopStart)) {
          Text(text = "\"Font size (pt)\"", modifier = Modifier.wrapContentWidth().wrapContentHeight().padding(top = 20.0.dp).padding(start = 72.0.dp), color = Color.Unspecified, fontSize = 16.sp)
        }
        Box(modifier = Modifier.align(Alignment.TopStart)) {
          Row(modifier = Modifier.fillMaxWidth().wrapContentHeight().padding(bottom = 8.0.dp).padding(start = 62.0.dp).padding(top = 36.0.dp), verticalAlignment = Alignment.CenterVertically) {
            Column(modifier = Modifier.wrapContentHeight().weight(1.0f)) {
            }
            OutlinedTextField(value = "", onValueChange = {}, modifier = Modifier.size(48.0.dp, 48.0.dp).padding(end = 16.0.dp))
          }
        }
      }
      Box(modifier = Modifier.fillMaxWidth().wrapContentHeight()) {
        Box(modifier = Modifier.align(Alignment.CenterStart)) {
          Box(modifier = Modifier.size(24.0.dp, 24.0.dp).padding(start = 16.0.dp)) { /* TODO Image */ }
        }
        Box(modifier = Modifier.align(Alignment.TopStart)) {
          Text(text = "\"Use pubkey authentication\"", modifier = Modifier.wrapContentWidth().wrapContentHeight().padding(top = 20.0.dp).padding(start = 72.0.dp), color = Color.Unspecified, fontSize = 16.sp)
        }
        Box(modifier = Modifier.align(Alignment.BottomStart)) {
          Text(text = "Use any key", modifier = Modifier.wrapContentWidth().wrapContentHeight().padding(bottom = 20.0.dp).padding(start = 72.0.dp).padding(top = 36.0.dp), color = Color.Unspecified, fontSize = 14.sp)
        }
      }
      Box(modifier = Modifier.fillMaxWidth().wrapContentHeight()) {
        Box(modifier = Modifier.align(Alignment.CenterStart)) {
          Box(modifier = Modifier.size(24.0.dp, 24.0.dp).padding(start = 16.0.dp)) { /* TODO Image */ }
        }
        Box(modifier = Modifier.align(Alignment.TopStart)) {
          Text(text = "\"DEL Key\"", modifier = Modifier.wrapContentWidth().wrapContentHeight().padding(top = 20.0.dp).padding(start = 72.0.dp), color = Color.Unspecified, fontSize = 16.sp)
        }
        Box(modifier = Modifier.align(Alignment.BottomStart)) {
          Text(text = "Delete", modifier = Modifier.wrapContentWidth().wrapContentHeight().padding(bottom = 20.0.dp).padding(start = 72.0.dp).padding(top = 36.0.dp), color = Color.Unspecified, fontSize = 14.sp)
        }
      }
      Box(modifier = Modifier.fillMaxWidth().wrapContentHeight()) {
        Box(modifier = Modifier.align(Alignment.CenterStart)) {
          Box(modifier = Modifier.size(24.0.dp, 24.0.dp).padding(start = 16.0.dp)) { /* TODO Image */ }
        }
        Box(modifier = Modifier.align(Alignment.TopStart)) {
          Text(text = "\"Encoding\"", modifier = Modifier.wrapContentWidth().wrapContentHeight().padding(top = 20.0.dp).padding(start = 72.0.dp), color = Color.Unspecified, fontSize = 16.sp)
        }
        Box(modifier = Modifier.align(Alignment.BottomStart)) {
          Text(text = "UTF-8", modifier = Modifier.wrapContentWidth().wrapContentHeight().padding(bottom = 20.0.dp).padding(start = 72.0.dp).padding(top = 36.0.dp), color = Color.Unspecified, fontSize = 14.sp)
        }
      }
      Column(modifier = Modifier.fillMaxWidth().wrapContentHeight()) {
      }
      Column(modifier = Modifier.fillMaxWidth().wrapContentHeight()) {
      }
      Column(modifier = Modifier.fillMaxWidth().wrapContentHeight()) {
      }
      Column(modifier = Modifier.fillMaxWidth().wrapContentHeight()) {
      }
      Column(modifier = Modifier.fillMaxWidth().wrapContentHeight()) {
      }
      Box(modifier = Modifier.fillMaxWidth().wrapContentHeight()) {
        Box(modifier = Modifier.align(Alignment.CenterStart)) {
          Box(modifier = Modifier.size(24.0.dp, 24.0.dp).padding(start = 16.0.dp)) { /* TODO Image */ }
        }
        Box(modifier = Modifier.align(Alignment.TopStart)) {
          Text(text = "\"Post-login automation\"", modifier = Modifier.wrapContentWidth().wrapContentHeight().padding(top = 20.0.dp).padding(start = 72.0.dp), color = Color.Unspecified, fontSize = 16.sp)
        }
        Box(modifier = Modifier.align(Alignment.TopStart)) {
          Text(text = "\"Commands to run on remote server once authenticated\"", modifier = Modifier.wrapContentWidth().wrapContentHeight().padding(bottom = 2.0.dp).padding(start = 72.0.dp).padding(top = 36.0.dp), color = Color.Unspecified, fontSize = 14.sp)
        }
        Box(modifier = Modifier.align(Alignment.TopStart)) {
          OutlinedTextField(value = "", onValueChange = {}, modifier = Modifier.fillMaxWidth().wrapContentHeight().padding(start = 72.0.dp).padding(top = 16.0.dp))
        }
      }
    }
  }
}


@Preview(showBackground = true)
@Composable
fun fragment_host_editor_preview() {
    MyApplicationTheme {
        fragment_host_editor()
    }
}

File: experiments/tools/connectbot/swiftui/Fragment_host_editor.swift
import SwiftUI

struct Fragment_host_editor: View {
  var body: some View {
    ScrollView {
      VStack(alignment: .leading, spacing: 0) {
        ZStack {
          Image(systemName: "photo").resizable().scaledToFit().frame(width: 24.0, height: 24.0).padding(.leading, 16.0).frame(maxWidth: .infinity, alignment: .leading)
          Text("Protocol").font(.system(size: 16)).foregroundColor(Color.clear).padding(.top, 20.0).padding(.leading, 72.0).frame(maxWidth: .infinity, maxHeight: .infinity, alignment: .topLeading)
          Text("ssh").font(.system(size: 14)).foregroundColor(Color.clear).padding(.bottom, 20.0).padding(.leading, 72.0).padding(.top, 36.0).frame(maxWidth: .infinity, maxHeight: .infinity, alignment: .bottomLeading)
        }
        .frame(maxWidth: .infinity)
        HStack(alignment: .center, spacing: 0) {
          VStack(alignment: .leading, spacing: 8) {
            /* TODO: TextInputLayout label */
            TextField("", text: .constant("")).frame(maxWidth: .infinity)
          }
          .frame(maxWidth: .infinity)
          Group {
          Button(action: { /* TODO */ }) {
            Image("ic_expand_more")
          }
          .background(Color(red: 0, green: 0, blue: 0, opacity: 0))
          .frame(width: 48.0, height: 48.0)
          .padding(16.0)
          }
          .frame(maxHeight: .infinity, alignment: .center)
        }
        .frame(maxWidth: .infinity)
        .padding(.leading, 68.0)
        ZStack {
          Image(systemName: "photo").resizable().scaledToFit().frame(width: 24.0, height: 24.0).padding(.leading, 16.0).frame(maxWidth: .infinity, alignment: .leading)
          VStack(alignment: .leading, spacing: 8) {
            /* TODO: TextInputLayout label */
            TextField("\"Nickname\"", text: .constant("")).frame(maxWidth: .infinity)
          }
          .frame(maxWidth: .infinity)
          .padding(.leading, 68.0)
          .padding(.trailing, 16.0)
        }
        .frame(maxWidth: .infinity)
        ZStack {
          Image(systemName: "photo").resizable().scaledToFit().frame(width: 24.0, height: 24.0).padding(.leading, 16.0).frame(maxWidth: .infinity, alignment: .leading)
          Text("\"Color category\"").font(.system(size: 16)).foregroundColor(Color.clear).padding(.top, 20.0).padding(.leading, 72.0).frame(maxWidth: .infinity, maxHeight: .infinity, alignment: .topLeading)
          Text("gray").font(.system(size: 14)).foregroundColor(Color.clear).padding(.bottom, 20.0).padding(.leading, 72.0).padding(.top, 36.0).frame(maxWidth: .infinity, maxHeight: .infinity, alignment: .bottomLeading)
        }
        .frame(maxWidth: .infinity)
        ZStack {
          Image(systemName: "photo").resizable().scaledToFit().frame(width: 24.0, height: 24.0).padding(.leading, 16.0).frame(maxWidth: .infinity, alignment: .leading)
          Text("\"Font size (pt)\"").font(.system(size: 16)).foregroundColor(Color.clear).padding(.top, 20.0).padding(.leading, 72.0).frame(maxWidth: .infinity, maxHeight: .infinity, alignment: .topLeading)
          Group {
          HStack(alignment: .center, spacing: 0) {
            Group {
            VStack(alignment: .leading, spacing: 0) {
            }
            .frame(maxWidth: .infinity)
            }
            .frame(maxHeight: .infinity, alignment: .center)
            TextField("", text: .constant("")).frame(width: 48.0, height: 48.0).padding(.trailing, 16.0).frame(maxHeight: .infinity, alignment: .center)
          }
          .frame(maxWidth: .infinity)
          .padding(.bottom, 8.0)
          .padding(.leading, 62.0)
          .padding(.top, 36.0)
          }
          .frame(maxWidth: .infinity, maxHeight: .infinity, alignment: .topLeading)
        }
        .frame(maxWidth: .infinity)
        ZStack {
          Image(systemName: "photo").resizable().scaledToFit().frame(width: 24.0, height: 24.0).padding(.leading, 16.0).frame(maxWidth: .infinity, alignment: .leading)
          Text("\"Use pubkey authentication\"").font(.system(size: 16)).foregroundColor(Color.clear).padding(.top, 20.0).padding(.leading, 72.0).frame(maxWidth: .infinity, maxHeight: .infinity, alignment: .topLeading)
          Text("Use any key").font(.system(size: 14)).foregroundColor(Color.clear).padding(.bottom, 20.0).padding(.leading, 72.0).padding(.top, 36.0).frame(maxWidth: .infinity, maxHeight: .infinity, alignment: .bottomLeading)
        }
        .frame(maxWidth: .infinity)
        ZStack {
          Image(systemName: "photo").resizable().scaledToFit().frame(width: 24.0, height: 24.0).padding(.leading, 16.0).frame(maxWidth: .infinity, alignment: .leading)
          Text("\"DEL Key\"").font(.system(size: 16)).foregroundColor(Color.clear).padding(.top, 20.0).padding(.leading, 72.0).frame(maxWidth: .infinity, maxHeight: .infinity, alignment: .topLeading)
          Text("Delete").font(.system(size: 14)).foregroundColor(Color.clear).padding(.bottom, 20.0).padding(.leading, 72.0).padding(.top, 36.0).frame(maxWidth: .infinity, maxHeight: .infinity, alignment: .bottomLeading)
        }
        .frame(maxWidth: .infinity)
        ZStack {
          Image(systemName: "photo").resizable().scaledToFit().frame(width: 24.0, height: 24.0).padding(.leading, 16.0).frame(maxWidth: .infinity, alignment: .leading)
          Text("\"Encoding\"").font(.system(size: 16)).foregroundColor(Color.clear).padding(.top, 20.0).padding(.leading, 72.0).frame(maxWidth: .infinity, maxHeight: .infinity, alignment: .topLeading)
          Text("UTF-8").font(.system(size: 14)).foregroundColor(Color.clear).padding(.bottom, 20.0).padding(.leading, 72.0).padding(.top, 36.0).frame(maxWidth: .infinity, maxHeight: .infinity, alignment: .bottomLeading)
        }
        .frame(maxWidth: .infinity)
        VStack(alignment: .leading, spacing: 0) {
        }
        .frame(maxWidth: .infinity)
        VStack(alignment: .leading, spacing: 0) {
        }
        .frame(maxWidth: .infinity)
        VStack(alignment: .leading, spacing: 0) {
        }
        .frame(maxWidth: .infinity)
        VStack(alignment: .leading, spacing: 0) {
        }
        .frame(maxWidth: .infinity)
        VStack(alignment: .leading, spacing: 0) {
        }
        .frame(maxWidth: .infinity)
        ZStack {
          Image(systemName: "photo").resizable().scaledToFit().frame(width: 24.0, height: 24.0).padding(.leading, 16.0).frame(maxWidth: .infinity, alignment: .leading)
          Text("\"Post-login automation\"").font(.system(size: 16)).foregroundColor(Color.clear).padding(.top, 20.0).padding(.leading, 72.0).frame(maxWidth: .infinity, maxHeight: .infinity, alignment: .topLeading)
          Text("\"Commands to run on remote server once authenticated\"").font(.system(size: 14)).foregroundColor(Color.clear).padding(.bottom, 2.0).padding(.leading, 72.0).padding(.top, 36.0).frame(maxWidth: .infinity, maxHeight: .infinity, alignment: .topLeading)
          TextField("", text: .constant("")).frame(maxWidth: .infinity).padding(.leading, 72.0).padding(.top, 16.0).frame(maxWidth: .infinity, maxHeight: .infinity, alignment: .topLeading)
        }
        .frame(maxWidth: .infinity)
      }
      .frame(maxWidth: .infinity)
    }
    .frame(maxWidth: .infinity, maxHeight: .infinity)
  }
}

struct Fragment_host_editor_Previews: PreviewProvider {
  static var previews: some View {
    Fragment_host_editor()
  }
}
